File: Leetcode/187. Repeated DNA Sequences/repeated_DNA_sequences.py
class Solution:
    # using set and combination of all patterns
    def findRepeatedDnaSequences(self, s: str):
        sequence =set()
        repeated =set()
        for i in range(len(s)-9):
            cur_seq = s[i:i+10]
            if cur_seq in sequence:
                repeated.add(cur_seq)
            else:
                sequence.add(cur_seq)
        return [*repeated]
    
    # Using Rabin Karp Method
    def findRepeatedDnaSequences1(self, s: str):
        if len(s)<10:
            return []
        input_coded = []
        for i in s:
            if i=="A":
                input_coded.append(1)
            elif i=="C":
                input_coded.append(2)
            elif i=="G":
                input_coded.append(3)
            else:
                input_coded.append(4)
        len_s = len(s)
        len_p = 10
        hash_val = 0
        base = 4
        hashmap = {}
        res = []
        for i in range(10):
            hash_val = hash_val*base + input_coded[i]
        hashmap[hash_val]=1
        for i in range(1,len_s-9):
            hash_val = hash_val*base - input_coded[i-1]*(base**len_p) +input_coded[i+9]
            hashmap[hash_val] = hashmap.get(hash_val,0)+1
            if hashmap[hash_val] ==2:
                res.append(s[i:i+10])
        return res
    
    def findRepeatedDnaSequences2(self, s: str):
        n2d = {'A': 0, 'C': 1, 'G': 2, 'T': 3}
        candidates = set()
        duplicates = set()
        cur = 0
        
        for i in range(len(s)):
            # only keep at most 9 letters before the current letter
			# or maybe cur &= (1<<18) -1 and extra first char is removed and space is maintained to accomadate new char
            cur &= (1<<18)-1
            #cur %= 1<<18

            #we need to left shift by 2 which can be acheieved by mutlying by 4 as 1 shift  = mutiple by 2
            cur = cur * 4 + n2d[s[i]]
            # skipping untill we get length 10
            if i < 9:
                continue
            
            if cur in candidates:
                duplicates.add(s[i - 9:i + 1])
            else:
                candidates.add(cur)
        return list(duplicates)

sol =Solution()
print(sol.findRepeatedDnaSequences("AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT"))
print(sol.findRepeatedDnaSequences1("AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT"))
print(sol.findRepeatedDnaSequences2("AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT"))

File: Leetcode/187. Repeated DNA Sequences/repeated_DNA_sequences.rs
use std::collections::{HashSet, HashMap};

struct Solution{}

impl Solution {
    pub fn find_repeated_dna_sequences(s: String) -> Vec<String> {
        if s.len()<10{
            return vec![];
        }
        let mut sequence:HashSet<&str> = HashSet::new();
        let mut duplicates:HashSet<&str> = HashSet::new();
        for i in 0..s.len()-9{
            let curr_pattern = &s[i..i+10];
            if sequence.contains(curr_pattern){
                duplicates.insert(curr_pattern);
            } else {
                sequence.insert(curr_pattern);
            }
        }
        duplicates.into_iter().map(String::from).collect()
    }
    pub fn find_repeated_dna_sequences1(s: String) -> Vec<String> {
        if s.len()<10{
            return vec![];
        }
        let n2d:HashMap<char,u32> =[('A', 0), ('C', 1), ('G', 3), ('T', 2)].iter().cloned().collect();
        let mut count_map:HashMap<i32, i32> = HashMap::new();
        let mut encoded_input:Vec<i32> = Vec::new();
        for i in s.chars(){
            encoded_input.push(n2d[&i] as i32)
        }
        let mut cur_hash = 0;
        let mut res:Vec<String> = Vec::new();
        let base:i32 = 4;
        for i in 0..10{
            cur_hash = cur_hash * base + encoded_input[i];
        }
        count_map.insert(cur_hash.clone(), 1);
        for i in 1..(s.len()-9){
            cur_hash = cur_hash*base - encoded_input[i-1]*(base.pow(10))+encoded_input[i+9];
            if count_map.contains_key(&cur_hash) {
                count_map.insert(cur_hash.clone(), count_map.get(&cur_hash).unwrap()+1);
                if count_map.get(&cur_hash).unwrap().clone()== 2_i32{
                    res.push(s.chars().skip(i).take(10).collect())
                }
                
            } else {
                count_map.insert(cur_hash.clone(), 1);
            }
        }
        res
    }
    pub fn find_repeated_dna_sequences2(s: String) -> Vec<String> {
        let mut cur:u32 = 0;
        let n2d:HashMap<char,u32> =[('A', 0), ('C', 1), ('G', 3), ('T', 2)].iter().cloned().collect();
        let mut candidates:HashSet<u32> = HashSet::new();
        let mut duplicates:HashSet<String> = HashSet::new();
        for (ch,i) in s.chars().zip(0..){
            cur &=((1 as u32)<<18)-1;
            cur = cur * 4 + n2d[&ch];
            if i >= 9 {
                if candidates.contains(&cur) {
                    duplicates.insert(s.chars().skip(i as usize - 9).take(10).collect());
                } else {
                    candidates.insert(cur);
                }
            }
        }
        duplicates.into_iter().collect()
    }
}
fn main() {
    let res = Solution::find_repeated_dna_sequences("AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT".to_string());
    let res1 = Solution::find_repeated_dna_sequences1("AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT".to_string());
    let res2= Solution::find_repeated_dna_sequences2("AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT".to_string());
    println!("{:?}\n{:?}\n{:?}",res,res1,res2);
}
